clang/lib/Sema/SemaBoundsSafety.cpp

```diff
@@ -132,9 +132,20 @@ bool Sema::CheckCountedByAttrOnField(FieldDecl *FD, Expr *E, bool CountInBytes,
     // `BoundsSafetyCheckUseOfCountAttrPtr`
     //
     // * When the pointee type is always an incomplete type (e.g.
-    // `void`) the attribute is disallowed by this method because we know the
-    // type can never be completed so there's no reason to allow it.
-    InvalidTypeKind = CountedByInvalidPointeeTypeKind::INCOMPLETE;
+    // `void` in strict C mode) the attribute is disallowed by this method
+    // because we know the type can never be completed so there's no reason
+    // to allow it.
+    //
+    // Exception: In GNU mode, void has an implicit size of 1 byte for pointer
+    // arithmetic. Therefore, counted_by on void* is allowed as a GNU extension
+    // and behaves equivalently to sized_by (treating the count as bytes).
+    bool IsVoidPtrInGNUMode = PointeeTy->isVoidType() && getLangOpts().GNUMode;
+    if (IsVoidPtrInGNUMode) {
+      // Emit a warning that this is a GNU extension
+      Diag(FD->getBeginLoc(), diag::ext_gnu_counted_by_void_ptr) << Kind;
```

Comment:
We should emit a note diagnostic here that suggests using `__sized_by` (or `__sized_by_or_null`) to remove this warning. Ideally we should add a fix-it to that too but that can be quite painful to implement because we don't have standardized way of writing the attributes (e.g. raw GNU attribute vs macro).

Reply:
I'm not suggesting you modify the wording here. I'm suggesting you **also** emit a note diagnostic that tells the user how to suppress the warning by using the appropriate attribute instead of `__counted_by`.